src/v/rpc/transport.cc

```diff
@@ -123,7 +123,14 @@ base_transport::connect(clock_type::time_point connection_timeout) {
 }
 ss::future<> base_transport::stop() {
     fail_outstanding_futures();
-    return _dispatch_gate.close();
+
+    // We must call stop() on our output stream, because seastar::output_stream
+    // may not be safely destroyed without a call to close(), and this class
+    // may be destroyed after stop() is called.
+    return _out.stop().then([this]() {
+        _out = {};
+        return _dispatch_gate.close();
```

Comment:
can you comment on why we close the gate after stopping the output? by far the most common pattern is to close the gate to prevent new fibers from entering, and then do clean-up after. but sometimes there are exceptions.

Reply:
Hmm, I think this is an oversight on my part -- I was focused on just making sure we didn't proceed without the stop.  I'll re-test with these flipped.

Reply:
there's a couple patterns that are common. my hunch is that in this case the first works, but usually requires looking a bit how the gate is used and the potential sources of new fibers.

```
return gate.close().then([] { ... });
```

and

```
auto f = gate.close();
...
return f;
```

or

```
auto f = gate.close();
return other.then([f=std::move(f)] { ...; return f; });
```